src/computation/logic/distribute-levelling-seats.ts

```diff
@@ -90,6 +65,164 @@ export function distributeLevelingSeats(
         );
     }
 
-    const levelingSeatDistribution = dictionaryToArray(partyRestQuotients);
-    return levelingSeatDistribution;
+    const levelingSeatDistributionArray = dictionaryToArray(partyRestQuotients);
+    return levelingSeatDistributionArray;
+}
+
+/**
+ * Filters out parties that do not receive more seats in the national distribution than the district distribution.
+ * It returns the remaining party codes and the final national distribution of seats.
+ *
+ * @param levelingPartyCodes The party codes to filter
+ * @param payload The computation payload
+ * @param partyResults The party results for each party
+ */
+function nationalDistributionFilter(
+    levelingPartyCodes: string[],
+    payload: ComputationPayload,
+    partyResults: Dictionary<PartyResult>
+): { nationalDistribution: DistributionResult; levelingPartyCodes: string[] } {
+    let localLevelingPartyCodes = [...levelingPartyCodes];
+    let nationalDistribution: DistributionResult;
+
+    let unfilteredPartiesLength;
+    let filteredPartiesLength = localLevelingPartyCodes.length;
+    do {
+        unfilteredPartiesLength = filteredPartiesLength;
+
+        let totalVotes = 0;
+        let seatsToDistribute = payload.levelingSeats;
+        const levelingParties: Result[] = [];
+        for (const partyCode of localLevelingPartyCodes) {
+            totalVotes += partyResults[partyCode].votes;
+            seatsToDistribute += partyResults[partyCode].districtSeats;
+            const party: Result = {
+                countyId: -1,
+                electionId: -1,
+                partyId: -1,
+                resultId: -1,
+                countyName: "",
+                partyCode,
+                partyName: "",
+                votes: partyResults[partyCode].votes,
+                percentage: -1,
+            };
+            levelingParties.push(party);
+        }
```

Comment:
For a later time, I think this could probably be achieved by a few rewrites and a map of some kind (assuming it doesn't harshly affect performance).

Reply:
I had a look at it. The issue is that Dictionary is not iterable, so we cannot easily use map or reduce without first converting it to an array.